### ISSUE TITLE
Do we need to insert bos and eos tokens fully in the train.csv?

### ISSUE THREAD

**Issue by jackswl** [huggingface/autotrain-advanced#753]
When fine-tuning LLM using train.csv, does the sample require the full template which includes the **bos and eos?**

For example, if the model bos_token is ```<s>```, do I need to include it into the train.csv sample as well?

**Comment by jackswl** [comment]
@abhishekkrthakur, for example, for mistral 7b instruct v0.3,
will each of the example in train.csv look something like this in the text column:
```<s>[INST] hi this is user[/INST] this is assistant </s>```

is this right? it will include the entire

**Comment by abhishekkrthakur** [comment]
you dont need to format dataset and keep it in json format if using chat template. for example no_robots dataset. if you dataset is plain test and you are training a chat model, you need to add special tokens and tags. there is a parameter that can add end token by the way.

**Comment by jackswl** [comment]
@abhishekkrthakur, i am currently using plain text. However, my plain text already contains all the special tokens and tags from pre-applying the chat template myself. Is this approach ok?

so for example, one of my sample inside train.csv can be:
```<s>[INST] hi this is user[/INST] this is assistant </s>```, which is already applied with the chat template beforehand. Just double confirming that this is ok for autotrain? So my train.csv will contains all the plain text in text column that already applied with chat template.

**Comment by abhishekkrthakur** [comment]
yes. in that case, make sure `chat_template` is set to none.

**Comment by jackswl** [comment]
ok thanks for the swift reply. was trying to re-confirm, so I don't get errors like having **EOS token applying again** on top of my train.csv during fine-tuning.

**Comment by jackswl** [comment]
@abhishekkrthakur for example:

messages = '''<s>[INST] You are an expert Python programmer. Your task is to do this.[/INST]
assistant goes here
</s>'''

tokenizer(messages) <-- this will output <s> again

By doing this, the <s> will be applied again. So I was wondering if autotrain will do this for my plain text? or should my plain text include all special tags EXCEPT **BOS token**?

**Comment by abhishekkrthakur** [comment]
choose llm generic and disable option to add end token and it will be fine.

**Comment by jackswl** [comment]
@abhishekkrthakur Could you help on this? Just re-clarifying. This is my original .yml file below. Do I change the task from ```llm-sft``` to ```llm```, then insert ```add_eos_token: false``` under ```data```, then also add ```trainer: default``` under ```data```? Did I leave out anything?

```
task: llm-sft
base_model: /scratch/xxx
project_name: xxx
log: none
backend: local

data:
  path: /home/xxx
  train_split: train
  valid_split: null
  chat_template: null
  column_mapping:
    text_column: text

params:
  block_size: 4096
  model_max_length: 4096
  epochs: 4
  batch_size: 4 
  lr: 1e-4 
  peft: true
  quantization: int4
  target_modules: all-linear
  padding: right
  optimizer: adamw_torch
  scheduler: cosine
  merge_adapter: false
  use_flash_attention_2: true   
  logging_steps: 1
  unsloth: false
  seed: 42
```

**Comment by jackswl** [comment]
@abhishekkrthakur because there is no generic template configs, hence I am not 100% sure on this. would be really beneficial if you can clarify on this.

**Comment by abhishekkrthakur** [comment]
its just llm. if you remove the sft and use only llm or add trainer: default, its generic training.

**Comment by abhishekkrthakur** [comment]
ill add a config :) thanks for letting me know

**Comment by jackswl** [comment]
> its just llm. if you remove the sft and use only llm or add trainer: default, its generic training.

sorry, you meant, i can simply just change the task from llm-sft to llm, and it should be ok? I believe i also need to add ```add_eos_token: false``` under data section is that right?

side note: does it mean if I use llm-sft as per original, then I do not have to add the BOS token, since it will be applied?

**Comment by jackswl** [comment]
@abhishekkrthakur and despite setting add_eos_token to be false under data section in .yml file, it still shows up as True

<img width="590" alt="image" src="https://github.com/user-attachments/assets/60e4ac0c-9e07-4bdc-a15e-73c9189e06e4">

**Comment by jackswl** [comment]
@abhishekkrthakur would greatly appreciate your reply on this, thanks a ton

**Comment by abhishekkrthakur** [comment]
ill add an example for your use case and update here asap :)

**Comment by jackswl** [comment]
@abhishekkrthakur thanks, I hope my problem was clear. 

I have a plain text (just a string) that is already applied with chat template. Which means it will include all the special tokens and tags. Things like EOS and BOS tokens will be applied as well already within this plain text.

I do not wish to have a duplicate BOS token (or EOS token) applying during fine tuning process using autotrain, because the tokenizer will add automatically append BOS token during fine tuning.

Wondering if generic llm like you mentioned can tackle this problem. 

Thanks for the help! Looking forward to the update 👍🏻🙏🏻

**Comment by jackswl** [comment]
@abhishekkrthakur any updates on this? thanks sir 👍🏻

**Comment by jackswl** [comment]
@abhishekkrthakur do you have any updates on this? I just want to ensure that I am doing the correct approach using your awesome package, for my use case (plain text in train.csv already has the full chat template applied). Thank you so much

**Comment by abhishekkrthakur** [comment]
unfortunately, i didnt get a chance to look deeper into it yet. but i will do it and update here as soon as possible. thank you for your patience.

**Comment by jackswl** [comment]
@abhishekkrthakur any updates on this? :(

**Comment by jackswl** [comment]
@abhishekkrthakur hi, any updates on this? thanks

**Comment by abhishekkrthakur** [comment]
hi. the best way as of now is to use SFT Trainer without the chat template. You can have all the data in the `text` column and you can format it the way you want. The same works for generic trainer. you have an extra option of `add_eos_token` in generic trainer which you can set to false in case you are formatting the data yourself.

**Comment by jackswl** [comment]
@abhishekkrthakur, you did not answer my question. let's say i am using generic trainer and set add_eos_token to be false.

so i assume I will add all the data in the text column **except for the <s>, right?** because the tokenizer will add BOS token automatically.

**Comment by abhishekkrthakur** [comment]
only relevant code is here: https://github.com/huggingface/autotrain-advanced/blob/a88c2f0574df497fe5367d584aba113ff0c52ed2/src/autotrain/trainers/clm/train_clm_default.py#L19
The tokenizer is called like this: https://github.com/huggingface/autotrain-advanced/blob/a88c2f0574df497fe5367d584aba113ff0c52ed2/src/autotrain/trainers/clm/utils.py#L165

**Comment by jackswl** [comment]
hi @abhishekkrthakur, thanks for the reply. i understand you might be swarmed with work, so it is inevitable to miss out the question - so let me be explicit again:

i am NOT talking about eos token. I am talking about **BOS token**, let's say, `<s>` in my case. When the script tokenize my samples in the text column, it will automatically have the BOS token appended in FRONT.

as such, the samples in the text column should have EVERYTHING, except for the BOS token. <-- this is what I am clarifying. would appreciate if you can let me know if that statement is true/false.

in other words, my text column should be:
messages = '''[INST] You are an expert Python programmer. Your task is to do this.[/INST]
assistant goes here
</s>'''

and NOT:
messages = '''<s>[INST] You are an expert Python programmer. Your task is to do this.[/INST]
assistant goes here
</s>'''

right?

**Comment by abhishekkrthakur** [comment]
No. The tokenizer wont add the BOS token. Apologies for the overlook

**Comment by jackswl** [comment]
@abhishekkrthakur, thanks for your reply. 

hence i am wondering why is that/where did it explicitly state that BOS token will not be added?

because in the utils.py script:
```
def tokenize(examples, tokenizer, config):
    output = tokenizer(examples[config.text_column])
    return output
```

when the examples in the text column are tokenized, my model's tokenizer will append BOS token to it automatically. I have tested this, and my string will have `<s>` appended to it when using the tokenizer.

**Comment by abhishekkrthakur** [comment]
```
In [11]: tok.decode([21017, 5524, 25, 1680, 345, 3551, 257, 1790, 9793, 546, 262, 23082, 286, 262, 3381, 366, 2144, 404, 1559, 88, 1, 287, 12446, 30, 4222, 779, 6096, 3519, 284, 2785, 158
    ...: 48, 1559, 444, 287, 262, 10515, 1910, 290, 21729, 5981, 2267, 13, 21017, 15286, 25, 366, 9069, 404, 1559, 88, 1, 10229, 284, 257, 1910, 4645, 810, 612, 318, 691, 530, 17872, 329,
    ...:  257, 1948, 922, 393, 2139, 13, 554, 12446, 11, 428, 3381, 318, 3573, 5981, 287, 262, 4827, 1910, 11, 810, 257, 15848, 1559, 88, 9749, 468, 2383, 1176, 625, 262, 9400, 290, 1762,
    ...:  3403, 286, 511, 4409, 13, 383, 4931, 286, 257, 15848, 1559, 88, 460, 1255, 287, 2793, 9400, 290, 5322, 7184, 6443, 329, 3259, 11, 355, 262, 9749, 468, 1310, 15660, 284, 2620, 94
    ...: 00, 393, 2148, 1365, 1762, 3403, 13, 198, 198, 26446, 2267, 468, 5174, 2785, 15848, 1559, 444, 287, 11798, 884, 355, 6308, 290, 3049, 2057, 11, 810, 257, 1178, 1588, 2706, 1630,
    ...: 257, 2383, 6903, 286, 262, 1910, 357, 33, 452, 641, 1222, 14136, 2978, 11, 2211, 737, 554, 777, 11798, 11, 3259, 1690, 1986, 1877, 9400, 11, 3614, 4034, 11, 290, 5322, 23189, 117
    ...: 6, 11, 3756, 284, 257, 3074, 810, 484, 389, 10795, 319, 262, 9749, 329, 511, 30489, 13, 770, 21403, 460, 1255, 287, 2252, 22711, 286, 9400, 290, 257, 7794, 287, 1762, 3403, 13, 1
    ...: 98, 198, 16350, 11, 262, 3721, 286, 15848, 1559, 88, 318, 6393, 284, 4547, 262, 17262, 286, 4827, 5939, 290, 262, 2928, 286, 1910, 1176, 319, 3259, 13, 7735, 2267, 318, 2622, 284
    ...: , 1833, 262, 6287, 290, 2928, 286, 15848, 1559, 444, 319, 262, 3773, 290, 284, 1205, 4788, 284, 2209, 428, 2071, 13, 198, 198, 19927, 25, 198, 33, 452, 641, 11, 449, 1539, 1222,
    ...: 14136, 2978, 11, 406, 13, 357, 6390, 737, 383, 7119, 286, 26040, 8393, 315, 1083, 290, 11302, 43793, 874, 355, 21259, 286, 371, 658, 287, 5849, 352, 22512, 554, 8988, 13, 4913, 2
    ...: 86, 11279, 29845, 1083, 11, 2681, 7, 18, 828, 7632, 12, 3695, 13, 21017, 5524, 25, 2735, 4727, 340, 284, 257, 3290, 50256])

Out[11]: '### Human: Can you write a short introduction about the relevance of the term "monopsony" in economics? Please use examples related to potential monopsonies in the labour market and cite relevant research.### Assistant: "Monopsony" refers to a market structure where there is only one buyer for a particular good or service. In economics, this term is particularly relevant in the labor market, where a monopsony employer has significant power over the wages and working conditions of their employees. The presence of a monopsony can result in lower wages and reduced employment opportunities for workers, as the employer has little incentive to increase wages or provide better working conditions.\n\nRecent research has identified potential monopsonies in industries such as retail and fast food, where a few large companies control a significant portion of the market (Bivens & Mishel, 2013). In these industries, workers often face low wages, limited benefits, and reduced bargaining power, leading to a situation where they are dependent on the employer for their livelihood. This dependence can result in further suppression of wages and a decline in working conditions.\n\nOverall, the concept of monopsony is essential to understanding the dynamics of labor markets and the impact of market power on workers. Further research is needed to understand the extent and impact of monopsonies on the economy and to develop policies to address this issue.\n\nReferences:\nBivens, J., & Mishel, L. (2013). The Pay of Corporate Executives and Financial Professionals as Evidence of Rents in Top 1 Percent Incomes. Journal of Economic Perspectives, 27(3), 57-78.### Human: Now explain it to a dog<|endoftext|>'
```

The input is the first sample from guanaco dataset. `add_eos_token` is set to true. as we can see, BOS token is not added by the tokenizer.

Another example:

```
In [12]: from transformers import AutoTokenizer

In [13]: tok = AutoTokenizer.from_pretrained("openai-community/gpt2")

In [14]: tok("hi, how are you?")
Out[14]: {'input_ids': [5303, 11, 703, 389, 345, 30], 'attention_mask': [1, 1, 1, 1, 1, 1]}

In [15]: tok.decode([5303, 11, 703, 389, 345, 30])
Out[15]: 'hi, how are you?'
```

As we can see, neither EOS nor BOS tokens are appended.

**Comment by jackswl** [comment]
<img width="797" alt="image" src="https://github.com/user-attachments/assets/bf2f5106-7572-4ff4-bab5-02aec0b8c4b5">

as per the image above, it is behaving differently. 
as such, should I then stick to whatever the model is giving?

**Comment by abhishekkrthakur** [comment]
i checked gpt2. let me check mistral and come back to you. it might be possible due to different configs

**Comment by abhishekkrthakur** [comment]
yes. it seems to be coming from the config:

https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/main/tokenizer_config.json#L2
<img width="521" alt="image" src="https://github.com/user-attachments/assets/80abb66b-2550-4eb3-abcd-dc0d37ae5068">


I can disable adding both EOS and BOS token in AutoTrain to make it consistent across all models. Does that work?

**Comment by jackswl** [comment]
@abhishekkrthakur, thank you so much for clarifying everything. this is clear now. in other words, when using certain models such as mistral, i should not include BOS token `<s>` in my text column, since it will be included when using their tokenizer in AutoTrain. (please correct me again if i am wrong here)

regarding AutoTrain, I believe:
1. if chat_template = null and user wishes to perform training using some kind of instructional template, they should have **pre-applied** the chat template to their dataset beforehand
2. this dataset pre-applied with chat template will then go into the text column for AutoTrain. this means that the samples in the text column will be included with EVERYTHING, including BOS, EOS, and whatever. the users should handle this fully themselves.
3. may be good to allow users to set both add_eos_token and add_bos_token under autotrain config, and not just add_eos_token

then again, i am not sure if disable adding both EOS and BOS token in autotrain will be better. these are just my 2 cents ^

**Comment by abhishekkrthakur** [comment]
yes you are right.

we have `add_eos_token` param to add the token only on sample end. 
i _think_ this is just a special case but ive to take a deeper look. i could not find `add_bos_token` in llama model for example: https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct/blob/main/tokenizer_config.json

so, added an add_bos_token and disabling addition by the model itself could work.
